shim/src/scopeHandler.js

```diff
@@ -1,128 +1,135 @@
-// safeEvaluator operation (the Proxy is the argument of a 'with' binding).
-// As described in createSafeEvaluator(), it has several functions:
-// * allow the very first (and only the very first) use of 'eval' to map to
-//   the real (unsafe) eval function, so it acts as a 'direct eval' and can
-//   access its lexical scope (which maps to the 'with' binding, which the
-//   ScopeHandler also controls)
-// * ensure that all subsequent uses of 'eval' map to the safeEvaluator,
-//   which lives as the 'eval' property of the safeGlobal
-// * route all other property lookups at the safeGlobal
-// * hide the unsafeGlobal which lives on the scope chain above the 'with'
-// * ensure the Proxy invariants despite some global properties being frozen
-
-import { getPrototypeOf, objectHasOwnProperty } from './commons';
-
-export class ScopeHandler {
-  // Properties stored on the handler are not available from the proxy.
-
-  // the Proxy is only used by with(), so the Handler only needs to implement
-  // a few properties: has, get, set (which we leave at the default)
-
-  // todo: throw if any traps other than get/set/has are run (e.g.
-  // getOwnPropertyDescriptors, apply, getPrototypeOf) . Make this handler
-  // inherit from a second one whose 'get' property always throws.
-
-  constructor(unsafeRec) {
-    this.unsafeGlobal = unsafeRec.unsafeGlobal;
-    this.unsafeEval = unsafeRec.unsafeEval;
-
-    // this flag allow us to determine if the eval() call is a controlled
+import { assign, create, freeze, getPrototypeOf, objectHasOwnProperty } from './commons';
+import { throwTantrum } from './utilities';
+
+/**
+ * alwayThrowHandler is a proxy handler which throws on any trap called.
+ * It's made from a proxy with a get trap that throws. Its target is
+ * an immutable (frozen) object and is safe to share.
+ */
+const alwayThrowHandler = new Proxy(freeze({}), {
+  get(target, prop) {
+    throwTantrum(`unexpected scope handler trap called: ${prop}`);
+  }
+});
+
+/**
+ * ScopeHandler manages a Proxy which serves as the global scope for the
+ * safeEvaluator operation (the Proxy is the argument of a 'with' binding).
+ * As described in createSafeEvaluator(), it has several functions:
+ * - allow the very first (and only the very first) use of 'eval' to map to
+ *   the real (unsafe) eval function, so it acts as a 'direct eval' and can
+ *    access its lexical scope (which maps to the 'with' binding, which the
+ *   ScopeHandler also controls).
+ * - ensure that all subsequent uses of 'eval' map to the safeEvaluator,
+ *   which lives as the 'eval' property of the safeGlobal.
+ * - route all other property lookups at the safeGlobal.
+ * - hide the unsafeGlobal which lives on the scope chain above the 'with'.
+ * - ensure the Proxy invariants despite some global properties being frozen.
+ */
+export function createScopeHandler(unsafeRec) {
+  // todo: this.shadowTarget = getPrototypeOf(somehow_get_target)
```

Comment:
What is this todo about?

Reply:
This was about keeping a reference to the shadow to avoid calling getPrototypeOf on the target every time the get trap is invoked.